common/unified/distributed/partition_kernels.cpp

```diff
@@ -0,0 +1,174 @@
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto i, auto mapping, auto range_starting_index) {
+            const auto prev_part = i > 0 ? mapping[i - 1] : comm_index_type{-1};
+            const auto cur_part = mapping[i];
+            range_starting_index[i] = cur_part != prev_part ? 1 : 0;
+        },
+        mapping.get_num_elems(), mapping, range_starting_index);
+    components::prefix_sum(exec, range_starting_index.get_data(),
+                           mapping.get_num_elems() + 1);
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto i, auto size, auto mapping,
+                      auto range_starting_index, auto ranges,
+                      auto range_parts) {
+            const auto prev_part = i > 0 ? mapping[i - 1] : comm_index_type{-1};
+            const auto cur_part = i < size ? mapping[i] : comm_index_type{-1};
+            if (cur_part != prev_part) {
+                auto out_idx = range_starting_index[i];
+                ranges[out_idx] = i;
+                if (i < size) {
+                    range_parts[out_idx] = cur_part;
+                }
+            }
+        },
+        mapping.get_num_elems() + 1, mapping.get_num_elems(), mapping,
+        range_starting_index, range_bounds, part_ids);
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_PARTITION_BUILD_FROM_MAPPING);
+
+
+template <typename GlobalIndexType>
+void build_ranges_from_global_size(std::shared_ptr<const DefaultExecutor> exec,
+                                   comm_index_type num_parts,
+                                   GlobalIndexType global_size,
+                                   Array<GlobalIndexType>& ranges)
+{
+    const auto size_per_part = global_size / num_parts;
+    const auto rest = global_size - (num_parts * size_per_part);
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto i, auto size_per_part, auto rest, auto ranges) {
+            ranges[i] = size_per_part + (i < rest ? 1 : 0);
+        },
+        ranges.get_num_elems() - 1, size_per_part, rest, ranges.get_data());
+    components::prefix_sum(exec, ranges.get_data(), ranges.get_num_elems());
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_PARTITION_BUILD_FROM_GLOBAL_SIZE);
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+void has_ordered_parts(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const distributed::Partition<LocalIndexType, GlobalIndexType>* partition,
+    bool* result)
+{
+    const auto part_ids = partition->get_part_ids();
+    const auto num_ranges = partition->get_num_ranges();
+    // it is necessary to use uint32 as a temporary result, since
+    // bool can't be used with suffles
+    Array<uint32> result_uint32{exec, 1};
+    run_kernel_reduction(
+        exec,
+        [] GKO_KERNEL(auto i, const auto part_ids) {
+            return static_cast<uint32>(part_ids[i] < part_ids[i + 1]);
+        },
+        [] GKO_KERNEL(const auto a, const auto b) { return a && b; },
```

Comment:
I think `&&` might be better here, because we really mean a full AND here, not a bitwise AND. The bitwise AND should also work, but then we are relying on the cast to always put the digit 1 in the same position while casting from `true`. I'm not sure what the standard says, though that's probably always the case. But still, I doubt it matters for performance and `&&` is clearer here.

Reply:
I guess you mean from logical point?
each entry is true/false -> 1/0 and we reduce them here. They are same meaning if they always 1/0.
If to keep the logical point, I would suggest we have explicit cast here.

Reply:
In which direction do mean the cast? For me, the only reasonable is casting the result to `uint32`. Also, I would like to stay with `&&` since it's clearer, as @Slaedr mentioned.

Reply:
Even though it's not explicitly specified (I will take care of that though!), run_kernel_reduction should internally assign every result to the ValueType deduced for the reduction, i.e. the identity element parameter. So we don't need to worry about returning a different type here.

Reply:
I've added the cast to `unit32`, although I don't think it should matter too much.